### DISCUSSION TITLE
Public crates registry

### DISCUSSION THREAD

**Discussion by ghost** [kellnr/kellnr#277] [General]
Hi,

I was thinking of hosting kellnr as an experimental alternative to crates.io. My biggest gripe with crates.io are the lack of namespaced packages and forcing the use of github. I see kellnr doesn't have user registration nor namespaced packages, so before requesting them, are there any plans to make this available for public instances?

P.S this should probably be a discussion, but the discussion tab isn't available.

**Comment by syphar** [comment]
just fyi, 

the [RFC for namespaces on crates.io & cargo is accepted](https://github.com/rust-lang/rfcs/pull/3243), and we're actively working on figuring out the details & implementing them, which of course might take some more time.

**Comment by syphar** [comment]
also, see https://github.com/rust-lang/crates.io/issues/326 , 

the crates.io team is not specifically against non-github login, but it's a big chunk of work to make it possible. 

So of course it could be easier in something like kellnr.

**Reply by ghost** [reply]
It looks like the project's architecture was conceived with github in mind, meaning a contribution to decouple crates.io from github that would make for enormous changes ([50+ files use github directly](https://github.com/search?q=repo%3Arust-lang%2Fcrates.io+github+language%3ARust&type=code&l=Rust)). Even the "simple" action of introducing a trait for an auth backend/client would immediately introduce dead code. Not to speak of reimplementing authentication interfaces in a github agnostic manner (one PR), then replacing the github client by that auth service/backend in 50+ files (that's one PR), writing a DB migration to convert all users to the new auth system (yet another PR), and maybe even more.

Such PRs have a low chance of exiting review hell in my experience and are best done by contributors who are either paid to do it as their job or contributors who have multiple months of nigh single focus on it.

IMO, it's easier to implement this in a new project or fork instead of trying to integrate it in an existing, tightly coupled legacy code base where maintainers have little to no time to review and integrate the changes. Code architecture is important and correcting it later can be herculean sometimes.

If someone does implement this in crates.io, I will applaud their tenacity as it is not something I possess for unpaid ventures.

**Comment by secana** [comment]
Hi,

a.t.m. I don't have the plan to make kellnr available as a public instance, as I do not want to create a competitor to crates.io but just an easy way to self-host crates that should not be public. If the RFC for namespaces is finalised, I'll check, if an implementation for kellnr makes sense, too.

**Reply by ghost** [reply]
So if contributions were made to allow hosting public instances you'd be against those? Just trying to get a picture, not demand you do. It's your project, so whatever you feel comfortable with (or not) is fine :slightly_smiling_face:

**Reply by secana** [reply]
Oh no, I would be super happy if someone contributes such changes. But when I wrote Kellnr, I did not optimize it for scaleability or performance but for ease of use. If you want to contribute code, to make Kellnr more usable as a public instance, I'm happy to merge it.

What I would not like to see is, that Kellnr gets used as an "anti crates.io", to break the crates eco-system into two parts and now everyone has to check if a crate is on crates.io or on Kellnr.
